### PR TITLE
CLOUDSTACK-10244: KVM online storage migration fails and corrupts disk

### PR DESCRIPTION
**Pull request by mike-tutkowski**
This is an issue that was introduced in the code-review process of https://github.com/apache/cloudstack/pull/2298.

What is probably the simplest solution (and what this PR does) is to store the result of MapUtils.isNotEmpty(mapMigrateStorage) before passing mapMigrateStorage to the replaceStorage method.

The replaceStorage method can mutate the data in the passed-in map, thus leading to a different result for MapUtils.isNotEmpty(mapMigrateStorage) later on (and causing an issue for online storage migration).

If we just store the value up front and refer to it later, we, of course, get the same (boolean) result and online storage migration works properly.

A more long-term solution (and one that I plan on implementing for 4.12) would be to make a shallow copy of the map inside of replaceStorage so that replaceStorage does not actually mutate the data in the original map. While that isn't a difficult solution to implement, I feel the one I have implemented in this PR is safer as we are in the release-candidate process already.

Here is a link to the JIRA ticket:

https://issues.apache.org/jira/browse/CLOUDSTACK-10244

I have tested the code and validated that it works.

This use case only applies to online storage migration from NFS or Ceph to managed storage, which is a new feature for 4.11.

### CODE REVIEW

**Comment by rhtyd** [comment]
@mike-tutkowski can you change the base of the PR to 4.11 branch?

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 1]
```diff
@@ -132,8 +132,9 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
             vmsnapshots = libvirtComputingResource.cleanVMSnapshotMetadata(dm);
 
             Map<String, MigrateCommand.MigrateDiskInfo> mapMigrateStorage = command.getMigrateStorage();
+            boolean migrateStorage = MapUtils.isNotEmpty(mapMigrateStorage);
```
Comment (wido) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: Would it not be safer to make this boolean a final one? That way you are 100% sure it's not changed anywhere.

Comment (mike-tutkowski) [reply plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java]: @wido Sounds good. I can do that and then put the code on the branch that @rhtyd asked.

**Comment by mike-tutkowski** [comment]
@rhtyd Will I need to open a new PR to change the base of the PR? Thanks!

**Comment by rhtyd** [comment]
@mike-tutkowski can you try the edit button on this pr?

**Comment by mike-tutkowski** [comment]
@rhtyd @wido It looks like it took another commit along for the ride when I put this PR on top of a new base. I'll just go ahead and open a new PR and comment on its number here.

**Comment by mike-tutkowski** [comment]
Closing this PR in favor of https://github.com/apache/cloudstack/pull/2416

**Pull request closed without merging**